walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,135 @@
 
 package com.google.sps;
 
+import java.lang.Math;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import com.google.sps.TimeRange;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    /* Takes a meeting request that has mandatory attendees, optional attendees, and a duration
+     * then returns a sorted Collection of all possible TimeRanges of when the meeting could take place.
+     * 
+     * This function works by first initializing a list of free blocks of a certain length
+     * Then it goes through events and removes free blocks when they overlap with an event.
+     * It then merges the remaining consecutive free blocks so they're outputted as one
+     * longer continuous TimeRange and we can verify that they're at least the length of the
+     * requested duration. Note that the function first tries to include both optional and mandatory
+     * attendees, but if no time ranges are found, only mandatory attendees are considered (if any).
+     */
+
+    // Get request information on attendees and duration
+    Collection<String> allAttendees = new ArrayList<String>(request.getAttendees());
+    allAttendees.addAll(request.getOptionalAttendees());
+    Collection<String> mandatoryAttendees = new ArrayList<String>(request.getAttendees());
+    long meetingDuration = request.getDuration();
+    
+    // First attempt to include both mandatory and optional attendees
+    ArrayList<TimeRange> freeTimes = new ArrayList<TimeRange>();
+    Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+
+    // Initialize all free time blocks. Size 30 refers to the length of each block in minutes
+    // (ie. blocks are 12:00am, 12:30am, 1:00am etc) since we assume all events begin on the half hour
+    // Note that 15, 10, 5, 2, 1 minutes will all work but the granularity is not necessary
+    populateFreeTimes(freeTimes, 30);
+
+    // Remove from freeTimes all the TimeRanges where an attendee has an event
+    removeBusyTimes(freeTimes, events, allAttendees, 30);
+
+    // Make the remaining free blocks continuous and verify they are of minimum duration
+    updateMeetingTimes(meetingTimes, freeTimes, meetingDuration);
+    
+    // Verify that there is at least one potential meeting time with all attendees
+    // If not, only consider mandatory attendees (if any)
+    if (meetingTimes.size() != 0) {
+      return meetingTimes;
+    }
+
+    // Consider only mandatory attendees
+    if (mandatoryAttendees.size() != 0) {
+      freeTimes.clear();
```

Comment:
Starting the whole process over will work, but is very inefficient. It would be good to think of a solution that allows you to generate the free time representation once to be used for both mandatory and optional case.